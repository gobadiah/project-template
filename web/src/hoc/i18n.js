import i18n from '~/config/i18n';

export default ({ req, namespaces }) => new Promise((resolve) => {
  if (req) {
    resolve(i18n.getInitialProps(req, namespaces));
  } else {
    resolve({});
  }
});
