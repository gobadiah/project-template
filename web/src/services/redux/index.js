import { createStore, combineReducers, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import thunkMiddleware from 'redux-thunk';
import { reducer as api } from 'redux-json-api';
import { isImmutable, fromJS } from 'immutable';
import { reducer as form } from 'redux-form';

import { reducer as auth } from '~/auth';

const middlewares = composeWithDevTools(applyMiddleware(thunkMiddleware));

const reducer = combineReducers({
  api,
  auth,
  form,
});

export default (state) => {
  // When re-creating state on client from serialized state from server,
  // we need to make immutable every component that want to be.
  if (state && state.auth && !isImmutable(state.auth)) {
    state.auth = fromJS(state.auth); // eslint-disable-line no-param-reassign
  }
  return createStore(reducer, state, middlewares);
};

export { default as withRedux } from 'next-redux-wrapper';
