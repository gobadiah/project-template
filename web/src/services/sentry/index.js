import { config, getRaven } from './config';

export { default as Sentry } from './tag';

export const capture = err => (
  err &&
  config.sentry.DSN &&
  !config.dev &&
  getRaven().captureException(err)
);
