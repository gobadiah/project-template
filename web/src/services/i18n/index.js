import i18n from 'i18next';
import backend from 'i18next-xhr-backend';
import LanguageDetector from 'i18next-browser-languagedetector';

export const availableLanguages = ['fr', 'en'];
export const availableNamespaces = ['common', 'index'];

const detection = {
  order: ['querystring', 'cookie', 'localStorage', 'navigator', 'header'],

  lookupQuerystring: 'lng',
  lookupCookie: 'i18next',
  lookupLocalStorage: 'i18nextLng',

  caches: false,
};

const options = {
  fallbackLng: 'fr',
  load: 'languageOnly',

  ns: availableNamespaces,
  defaultNS: 'common',

  debug: false,

  react: {
    wait: true,
  },

  saveMissing: true,

  detection,

  interpolation: {
    escapeValue: false,
  },
};

if (process.browser) {
  i18n.use(backend)
    .use(LanguageDetector);
  options.caches = ['localStorage', 'cookie'];
  options.backend = {
    loadPath: '/locales/{{lng}}/{{ns}}.json',
    addPath: '/locales/add/{{lng}}/{{ns}}',
  };
}

if (!i18n.isInitialized) {
  i18n.init(options);
}

i18n.getInitialProps = (req, nss) => {
  let namespaces = nss;
  if (typeof namespaces === 'undefined') {
    namespaces = i18n.options.defaultNS;
  }
  if (typeof namespaces === 'string') {
    namespaces = [namespaces];
  }

  req.i18n.toJSON = () => null;

  const initialI18nStore = {};
  req.i18n.languages.forEach((l) => {
    initialI18nStore[l] = {};
    namespaces.forEach((ns) => {
      initialI18nStore[l][ns] = req.i18n.services.resourceStore.data[l] ?
        req.i18n.services.resourceStore.data[l][ns] || {} : {};
    });
  });

  return {
    i18n: req.i18n,
    initialI18nStore,
    initialLanguage: req.i18n.language,
  };
};

export default i18n;
