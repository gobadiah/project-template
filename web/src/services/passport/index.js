import passport from 'passport';
import { OAuth2Strategy as GoogleStrategy } from 'passport-google-oauth';
import { Strategy as FacebookStrategy } from 'passport-facebook';
import axios from 'axios';

import { config as axiosConfig } from '~/services/axios';

import config from './config';

const google = config.google.clientId && config.google.clientSecret;
const facebook = config.facebook.clientId && config.facebook.clientSecret;

const getConfig = provider => ({
  clientID: config[provider].clientId,
  clientSecret: config[provider].clientSecret,
  callbackURL: `${config.web}/auth/${provider}/callback`,
});

const callback = provider => (accessToken, refreshToken, profile, done) => {
  const data = {
    uid: profile.id,
    provider,
    access_token: accessToken,
  };
  return axios.post('/auth/provider', data, axiosConfig())
    .then(result => done(null, result.data.data))
    .catch(err => done(err));
};

if (google) {
  passport.use(new GoogleStrategy(
    getConfig('google'),
    callback('google'),
  ));
}

if (facebook) {
  passport.use(new FacebookStrategy(
    getConfig('facebook'),
    callback('facebook'),
  ));
}

const setupProvider = (server, provider) => {
  server.get(
    `/auth/${provider}`,
    passport.authenticate(
      provider,
      {
        session: false,
        scope: config[provider].scope,
      },
    ),
  );

  server.get(
    `/auth/${provider}/callback`,
    passport.authenticate(
      provider,
      {
        failureRedirect: '/signin',
        session: false,
      },
    ),
    (req, res) => res.redirect('/'),
  );
};

export default (server) => {
  if (!google && !facebook) {
    return;
  }
  server.use(passport.initialize());

  if (google) {
    setupProvider(server, 'google');
  }

  if (facebook) {
    setupProvider(server, 'facebook');
  }
};
