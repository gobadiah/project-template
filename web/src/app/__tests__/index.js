import path from 'path';
import request from 'supertest';

import createApp from '..';

describe('App integration', () => {
  let server;

  beforeAll(() => createApp(false).then((value) => { server = value.listen(0); }));

  afterAll(() => { server.close(); });

  it('should respond with 200 status for index and contain some french text', () =>
    request(server)
      .get('/')
      .expect(200)
      .expect(/>Salut le monde !<\/div>/));

  /**
    * @todo Mock the proxy call. It might not be straightforward.
  it('should response normally to non-next endpoint (e.g. /api)', () => {
    const mock = new MockAdapter(axios);
    mock.onGet('http://127.0.0.1:8000')
      .reply(200, 'Api endpoint');

    return request(server)
      .get('/api')
      .expect(200)
      .expect('Api endpoint');
  });
  */
});

describe('App unit', () => {
  it('should have options matching snapshot', () => {
    const { options } = require('..');
    expect(options.backend.loadPath).toEqual(path.join(
      __dirname,
      '../../../locales/{{lng}}/{{ns}}.json',
    ));
    expect(options.backend.addPath).toEqual(path.join(
      __dirname,
      '../../../locales/{{lng}}/{{ns}}.missing.json',
    ));
    delete options.backend.loadPath;
    delete options.backend.addPath;
    expect(options).toMatchSnapshot();
  });

  it('should use dependencies', () => {
    jest.resetModules();
    const availableLanguages = ['fr'];
    const availableNamespaces = ['common'];
    require('~/services/i18n').availableLanguages = availableLanguages;
    require('~/services/i18n').availableNamespaces = availableNamespaces;
    const i18n = require('~/services/i18n').default;
    i18n.use.mockImplementation(() => i18n);
    i18n.init.mockImplementation(() => {});

    const mockBackend = 'backend';

    jest.mock('../../services/i18n');
    jest.mock('i18next-node-fs-backend', () => mockBackend);

    const LanguageDetector = 'LanguageDetector';
    const i18nextMiddleware = require('i18next-express-middleware').default;
    const handle = 'handle';
    const missingKeyHandler = 'missingKeyHandler';
    i18nextMiddleware.handle.mockImplementation(() => handle);
    i18nextMiddleware.missingKeyHandler.mockImplementation(() => missingKeyHandler);
    require('i18next-express-middleware').LanguageDetector = LanguageDetector;

    jest.mock('i18next-express-middleware');

    jest.mock('serve-favicon');
    const favicon = require('serve-favicon');
    favicon.mockImplementation(() => 'favicon');

    const dev = false;

    const app = require('..').default;
    const { options } = require('..');
    const result = app(dev);
    expect(typeof result.then === 'function').toBe(true);

    expect(i18n.use).toHaveBeenCalledTimes(2);
    expect(i18n.use).toHaveBeenCalledWith(mockBackend);
    expect(i18n.use).toHaveBeenCalledWith(LanguageDetector);

    expect(i18n.init).toHaveBeenCalledTimes(1);
    expect(i18n.init.mock.calls[0][0]).toEqual(options);
    const func = i18n.init.mock.calls[0][1];

    const mockApp = {
      prepare: jest.fn(() => ({
        then: jest.fn(fn => fn()),
      })),
    };

    jest.mock('next', () => jest.fn(() => mockApp));

    const next = require('next');

    const mockHandler = 'handler';
    jest.mock('../../routes', () => ({
      getRequestHandler: jest.fn(() => mockHandler),
    }));

    const passport = require('~/services/passport').default;
    jest.mock('../../services/passport');
    passport.mockImplementation(() => {});

    const express = require('express');
    const mockStatic = 'static';
    express.static.mockImplementation(() => mockStatic);
    const mockServer = {
      use: jest.fn(),
      get: jest.fn(),
      post: jest.fn(),
    };
    express.mockImplementation(() => mockServer);
    jest.mock('express-http-proxy');

    jest.mock('express');

    func();

    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith({ dev, dir: './src' });

    expect(favicon).toHaveBeenCalledTimes(1);
    expect(favicon).toHaveBeenCalledWith(path.join(
      __dirname,
      '../../static',
      'favicon.ico',
    ));

    expect(i18nextMiddleware.handle).toHaveBeenCalledTimes(1);
    expect(i18nextMiddleware.handle).toHaveBeenCalledWith(i18n);

    expect(mockServer.use).toHaveBeenCalledTimes(4);
    expect(mockServer.use).toHaveBeenCalledWith('favicon');
    expect(mockServer.use).toHaveBeenCalledWith(handle);
    expect(mockServer.use).toHaveBeenCalledWith('/locales', mockStatic);
    // @todo test for proxy to be rightly called
    // expect(mockServer.use).toHaveBeenCalledWith('/api', 'proxy');

    // expect(mockProxy).toHaveBeenCalledTimes(1);
    // expect(mockProxy).toHaveBeenCalledWith('');

    expect(passport).toHaveBeenCalledTimes(1);
    expect(passport).toHaveBeenCalledWith(mockServer);

    expect(mockServer.get).toHaveBeenCalledTimes(2);
    expect(mockServer.get.mock.calls[0][0]).toEqual('/api');
    const func2 = mockServer.get.mock.calls[0][1];
    const res = {};
    res.status = jest.fn(() => res);
    res.send = jest.fn(() => res);
    res.end = jest.fn();

    func2(undefined, res);
    expect(res.status).toHaveBeenCalledTimes(1);
    expect(res.status).toHaveBeenCalledWith(200);
    expect(res.send).toHaveBeenCalledTimes(1);
    expect(res.send).toHaveBeenCalledWith('Api endpoint');
    expect(res.end).toHaveBeenCalledTimes(1);

    expect(mockServer.get).toHaveBeenCalledWith('*', mockHandler);

    expect(i18nextMiddleware.missingKeyHandler).toHaveBeenCalledTimes(1);
    expect(i18nextMiddleware.missingKeyHandler).toHaveBeenCalledWith(i18n);

    expect(mockServer.post).toHaveBeenCalledTimes(1);
    expect(mockServer.post).toHaveBeenCalledWith('/locales/add/:lng/:ns', missingKeyHandler);

    expect(result).resolves.toBe(mockServer);
  });
});
