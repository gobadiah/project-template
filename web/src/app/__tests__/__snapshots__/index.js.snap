// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App unit should have options matching snapshot 1`] = `
Object {
  "backend": Object {
    "jsonIndent": 2,
  },
  "ns": Array [
    "common",
    "index",
    "home",
    "account",
    "sign-in",
    "session",
    "video",
  ],
  "preload": Array [
    "fr",
    "en",
  ],
}
`;
